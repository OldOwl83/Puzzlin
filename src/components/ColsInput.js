import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
