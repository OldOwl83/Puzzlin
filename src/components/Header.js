import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
