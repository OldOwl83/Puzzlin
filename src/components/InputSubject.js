/*
* An input box to set the search keywords in the URL query parameters.
*/

import React, { useState } from "react";
import { useLocation, useNavigate } from "react-router-dom";

import '../styles/InputSubject.css';

export const InputSubject = () => {

    // console.log( 'InputSubject' );

    const navigate = useNavigate();
    const location = useLocation();

    const [ input, setInput ] = useState( '' );
    
    const submitInput = (e) => {
        e.preventDefault();
        
        setInput( '' );

        navigate( `../?q=${ input }`, { 
            //if the query is repeated immediately, the location is not duplicated in browser history.
            replace: location.search === `?q=${ input }` ? true : false,
     } );
    }

    return (

        <h2>Enter a subject and select a photo. Then, press on the preview to start.
            <form onSubmit={ submitInput }>
                <input 
                    id="inputSubject"
                    type="text"
                    autoComplete="on"
                    value={ input }
                    onChange={ (e) => setInput( e.target.value ) }
                />
            </form>
        </h2>
    )
}