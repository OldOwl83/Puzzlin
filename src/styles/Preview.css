#previewContainer{
    width: 70%;

    margin-right: 4%;
}

#imagePreview1, #imagePreview2{
    width: 100%;
    max-height: 420px;
    
    border: 6px double #555;
    transition: border 800ms;
}

#imagePreview1{
    animation: fadeIn1 700ms;
}

#imagePreview2{
    animation: fadeIn2 700ms;
}

@keyframes fadeIn1 {
    from { opacity: 0;}
    to { opacity: 1;}
}
@keyframes fadeIn2 {
    from { opacity: 0;}
    to { opacity: 1;}
}

#imagePreview1:hover, #imagePreview2:hover{
    border: 6px solid rgb(238, 238, 129);
}

#author{
    font-size: 0.8rem;
    text-align: left;
}