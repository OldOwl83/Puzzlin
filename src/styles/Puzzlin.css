@font-face
{
        font-family: 'Dongle';
        font-weight: normal;
        src: url('fonts/Dongle/Dongle-Regular.ttf');
}

@font-face
{
        font-family: 'Dongle';
        font-weight: bold;
        src: url('fonts/Dongle/Dongle-Bold.ttf');
}

@font-face
{
        font-family: 'Dongle';
        font-weight: 200;
        src: url('fonts/Dongle/Dongle-Light.ttf');
}

@import url('https://fonts.googleapis.com/css2?family=Coiny&display=swap');

header{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

#logoContainer{
    display: flex;
    align-items: center;
    justify-content: center;
    
    width: 20%;
    margin: 16px;
}

#logoContainer img{
    width: 60%;

    border: 3px solid rgb(230, 185, 127);
}

h1{
    width: 50%;
    margin: 0;

    color: rgb(1, 80, 83);
    font-size: 5rem;
    font-family: 'Coiny';
    
    text-align: center;
}

hr{
    border-color: beige;
}