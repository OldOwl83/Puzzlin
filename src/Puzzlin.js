import { 
    BrowserRouter as Router,
    Routes, Route,
        } from "react-router-dom"; 

import { createContext, useState } from "react";
import { PhotoSelector } from "./components/PhotoSelector";
import { Puzzle } from "./components/Puzzle";

import './styles/Puzzlin.css';
import { NoRoute } from "./components/NoRoute";

export const MainPhotoContext = createContext( null );
export const PuzzleGrid = createContext( [] );

export const Puzzlin = () => {

    const [ photoPuzzle, setPhotoPuzzle ] = useState( '' );
    const [ puzzleGrid, setPuzzleGrid ] = useState( [4, 3] );

    return (
        <>
            <h1>Puzzlin</h1>
            <hr />

            <Router>
                <MainPhotoContext.Provider value={ { photoPuzzle, setPhotoPuzzle } } >
                <PuzzleGrid.Provider value={ { puzzleGrid, setPuzzleGrid } }>

                    <Routes>

                        <Route path="/" element={ <PhotoSelector /> } />
                        <Route path="/?q=:query" element={ <PhotoSelector /> } />

                        <Route path="/puzzle" element={ <Puzzle /> } />

                        <Route path="/*" element={ <NoRoute /> } />

                    </Routes>

                </PuzzleGrid.Provider>
                </MainPhotoContext.Provider>
            </Router>

        </>
    )
};