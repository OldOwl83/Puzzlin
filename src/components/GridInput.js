/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
