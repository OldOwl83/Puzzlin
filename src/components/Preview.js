/*
* This component displays the selected photo stored in the location state. The displayed image has a link to the puzzle board and sets the main photo in context when the user clicks on it.
*/

import { Link, useLocation } from "react-router-dom";
import React, { useState, memo, useContext } from "react";

import '../styles/Preview.css';
import 'animate.css';

import { MainPhotoContext } from '../Puzzlin';

export const Preview = memo(() => {

    // console.log( 'Preview' );

    const { setPuzzlePhoto } = useContext( MainPhotoContext );
    const location = useLocation();
    const [fade, setFade] = useState();

    return (
        <>
        <Link to='/puzzle' id="photoPreviewContainer">

            <img 
                src={ location.state?.selectedPhoto } 
                alt="Preview" 
                draggable='false' 
                id={ fade ? "imagePreview1" : 'imagePreview2'}
                onClick={ ( e ) => setPuzzlePhoto( e.target.src ) }
                onLoad={ () => { setFade( !fade )} }
            />

        </Link>

        <footer id="author">Photo by { location.state?.author || 'Mauro Donnantuoni' }</footer>
        </>
    )
})