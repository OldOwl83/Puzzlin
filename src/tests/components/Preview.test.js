import 'react';
import { MemoryRouter } from 'react-router-dom';

import { mount } from 'enzyme';

import { Preview } from '../../components/Preview';
import { MainPhotoContext } from '../../Puzzlin';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
    useLocation: () => { 
        return { 
            key: 'dsgfdgs',
            pathname: '/',
            state: {
                selectedPhoto: 'portada.jpg',
                author: 'Napoleón'
            },
        }; },
}));

describe('Tests on <Preview /> component', () => {

    const setPhotoPuzzle = jest.fn();

    const wrapper = mount( 
        <MemoryRouter>
            <MainPhotoContext.Provider value={ { setPhotoPuzzle } }>
                <Preview /> 

            </MainPhotoContext.Provider>
        </MemoryRouter>
        );
  
    test('It should match the snapshot.', () => {
      
        expect( wrapper ).toMatchSnapshot();
    });

    test('The component elements should match the location state data', () => {
      
        expect( wrapper.find( 'img' ).prop( 'src' ) ).toBe( 'portada.jpg' );
        expect( wrapper.find( 'footer' ).text()).toBe( 'Photo by Napoleón' );
    });

    test('It should call "setPhotoPuzzle" function on click event with source image data.', () => {
      
        wrapper.find( 'img' ).simulate( 'click', { target: { src: 'portada.jpg' }} );

        expect( setPhotoPuzzle ).toHaveBeenCalledTimes( 1 );
        expect( setPhotoPuzzle ).toHaveBeenCalledWith( 'portada.jpg' );
    });
    
    
    
});
