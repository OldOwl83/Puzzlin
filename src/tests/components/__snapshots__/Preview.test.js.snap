// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <Preview /> component It should match the snapshot. 1`] = `
Array [
  <a
    href="/puzzle"
    id="photoPreviewContainer"
    onClick={[Function]}
  >
    <img
      alt="Preview"
      draggable="false"
      id="imagePreview2"
      onClick={[Function]}
      onLoad={[Function]}
      src="portada.jpg"
    />
  </a>,
  <footer
    id="author"
  >
    Photo by 
    Napole√≥n
  </footer>,
]
`;
